File: services/geospatial-intelligence-service/src/geospatial_base/domain/engine/clustering.rs
use std::collections::HashMap;

use crate::models::{
    feature::Coordinate,
    layer::LayerDefinition,
    spatial_index::{ClusterAlgorithm, ClusterRequest, ClusterResponse, ClusterSummary},
};

pub fn cluster(layer: &LayerDefinition, request: &ClusterRequest) -> ClusterResponse {
    match request.algorithm {
        ClusterAlgorithm::Dbscan => dbscan(layer, request.radius_km.unwrap_or(8.0)),
        ClusterAlgorithm::KMeans => kmeans(layer, request.cluster_count.unwrap_or(3).max(1)),
    }
}

fn dbscan(layer: &LayerDefinition, radius_km: f64) -> ClusterResponse {
    let grid_step = (radius_km / 111.0).max(0.05);
    let mut buckets = HashMap::<(i64, i64), Vec<Coordinate>>::new();
    for feature in &layer.features {
        let centroid = feature.geometry.centroid();
        let key = (
            (centroid.lat / grid_step).floor() as i64,
            (centroid.lon / grid_step).floor() as i64,
        );
        buckets.entry(key).or_default().push(centroid);
    }

    let mut outliers = 0usize;
    let mut clusters = Vec::new();
    for (cluster_id, (_bucket, members)) in buckets.into_iter().enumerate() {
        if members.len() < 2 {
            outliers += members.len();
            continue;
        }

        let centroid = average_coordinate(&members);
        clusters.push(ClusterSummary {
            cluster_id: format!("dbscan-{}", cluster_id + 1),
            centroid,
            member_count: members.len(),
            density: members.len() as f64 / radius_km.max(1.0),
        });
    }

    clusters.sort_by(|left, right| right.member_count.cmp(&left.member_count));
    ClusterResponse {
        algorithm: ClusterAlgorithm::Dbscan,
        clusters,
        outliers,
    }
}

fn kmeans(layer: &LayerDefinition, cluster_count: usize) -> ClusterResponse {
    let mut groups = vec![Vec::new(); cluster_count];
    for (index, feature) in layer.features.iter().enumerate() {
        groups[index % cluster_count].push(feature.geometry.centroid());
    }

    let clusters = groups
        .into_iter()
        .enumerate()
        .filter(|(_, members)| !members.is_empty())
        .map(|(index, members)| ClusterSummary {
            cluster_id: format!("kmeans-{}", index + 1),
            centroid: average_coordinate(&members),
            member_count: members.len(),
            density: members.len() as f64 / cluster_count as f64,
        })
        .collect();

    ClusterResponse {
        algorithm: ClusterAlgorithm::KMeans,
        clusters,
        outliers: 0,
    }
}

fn average_coordinate(points: &[Coordinate]) -> Coordinate {
    let (lat_sum, lon_sum) = points.iter().fold((0.0, 0.0), |acc, point| {
        (acc.0 + point.lat, acc.1 + point.lon)
    });
    Coordinate {
        lat: lat_sum / points.len() as f64,
        lon: lon_sum / points.len() as f64,
    }
}

File: services/ontology-exploratory-analysis-service/internal/domain/spatial/clustering.go
package spatial

import (
	"fmt"
	"math"
	"sort"

	"github.com/openfoundry/openfoundry-go/services/ontology-exploratory-analysis-service/internal/models"
)

// Default knobs match the Rust impl byte-for-byte.
const (
	defaultDBSCANRadiusKm  = 8.0
	defaultKMeansClusters  = 3
	dbscanGridStepDegPerKm = 111.0
	dbscanGridStepFloor    = 0.05
)

// Cluster mirrors `domain::engine::clustering::cluster` in Rust.
func Cluster(layer models.LayerDefinition, request models.ClusterRequest) models.ClusterResponse {
	switch request.Algorithm {
	case models.ClusterAlgorithmDBSCAN:
		return dbscan(layer, ptrFloatOr(request.RadiusKm, defaultDBSCANRadiusKm))
	case models.ClusterAlgorithmKMeans:
		count := ptrIntOr(request.ClusterCount, defaultKMeansClusters)
		if count < 1 {
			count = 1
		}
		return kmeans(layer, count)
	default:
		return models.ClusterResponse{Algorithm: request.Algorithm, Clusters: []models.ClusterSummary{}}
	}
}

type bucketKey struct {
	lat int64
	lon int64
}

func dbscan(layer models.LayerDefinition, radiusKm float64) models.ClusterResponse {
	gridStep := math.Max(radiusKm/dbscanGridStepDegPerKm, dbscanGridStepFloor)

	// Preserve insertion order so cluster numbering is deterministic
	// across runs — Rust uses HashMap which is non-deterministic, but
	// we round-trip through a stable order to keep tests reproducible
	// without changing observable cluster contents.
	order := make([]bucketKey, 0)
	buckets := make(map[bucketKey][]models.Coordinate)
	for _, f := range layer.Features {
		c := f.Geometry.Centroid()
		key := bucketKey{
			lat: int64(math.Floor(c.Lat / gridStep)),
			lon: int64(math.Floor(c.Lon / gridStep)),
		}
		if _, seen := buckets[key]; !seen {
			order = append(order, key)
		}
		buckets[key] = append(buckets[key], c)
	}

	outliers := 0
	clusters := make([]models.ClusterSummary, 0, len(order))
	clusterID := 0
	denom := math.Max(radiusKm, 1.0)
	for _, key := range order {
		members := buckets[key]
		if len(members) < 2 {
			outliers += len(members)
			continue
		}
		clusterID++
		clusters = append(clusters, models.ClusterSummary{
			ClusterID:   fmt.Sprintf("dbscan-%d", clusterID),
			Centroid:    averageCoordinate(members),
			MemberCount: len(members),
			Density:     float64(len(members)) / denom,
		})
	}

	sort.SliceStable(clusters, func(i, j int) bool {
		return clusters[i].MemberCount > clusters[j].MemberCount
	})

	return models.ClusterResponse{
		Algorithm: models.ClusterAlgorithmDBSCAN,
		Clusters:  clusters,
		Outliers:  outliers,
	}
}

func kmeans(layer models.LayerDefinition, clusterCount int) models.ClusterResponse {
	groups := make([][]models.Coordinate, clusterCount)
	for i, f := range layer.Features {
		idx := i % clusterCount
		groups[idx] = append(groups[idx], f.Geometry.Centroid())
	}

	clusters := make([]models.ClusterSummary, 0, clusterCount)
	for i, members := range groups {
		if len(members) == 0 {
			continue
		}
		clusters = append(clusters, models.ClusterSummary{
			ClusterID:   fmt.Sprintf("kmeans-%d", i+1),
			Centroid:    averageCoordinate(members),
			MemberCount: len(members),
			Density:     float64(len(members)) / float64(clusterCount),
		})
	}

	return models.ClusterResponse{
		Algorithm: models.ClusterAlgorithmKMeans,
		Clusters:  clusters,
		Outliers:  0,
	}
}

func averageCoordinate(points []models.Coordinate) models.Coordinate {
	if len(points) == 0 {
		return models.Coordinate{}
	}
	var latSum, lonSum float64
	for _, p := range points {
		latSum += p.Lat
		lonSum += p.Lon
	}
	n := float64(len(points))
	return models.Coordinate{Lat: latSum / n, Lon: lonSum / n}
}
